pytorch_toolkit/formula_recognition/im2latex/models/text_recognition_heads/attention_based.py

```diff
@@ -0,0 +1,280 @@
+        self.embedding = nn.Embedding(out_size, self.emb_size)
+
+        # enc_rnn_h*2 is the dimension of context
+        self.W_c = nn.Linear(self.dec_rnn_h+2*self.enc_rnn_h, self.enc_rnn_h)
+        self.W_out = nn.Linear(self.enc_rnn_h, out_size)
+
+        # a trainable initial hidden state V_h_0 for each row
+        self.V_h_0 = nn.Parameter(torch.Tensor(self.n_layer*2, self.enc_rnn_h))
+        self.V_c_0 = nn.Parameter(torch.Tensor(self.n_layer*2, self.enc_rnn_h))
+        init.uniform_(self.V_h_0, -INIT, INIT)
+        init.uniform_(self.V_c_0, -INIT, INIT)
+
+        # Attention mechanism
+        self.beta = nn.Parameter(torch.Tensor(self.dec_rnn_h))
+        init.uniform_(self.beta, -INIT, INIT)
+        self.W_h = nn.Linear(self.dec_rnn_h, self.dec_rnn_h)
+        self.W_v = nn.Linear(self.enc_rnn_h*2, self.dec_rnn_h)
+
+    def forward(self, features, formulas=None):
+        """args:
+        imgs: [B, C, H, W]
+        formulas: [B, MAX_LEN]
+
+        return:
+        logits: [B, MAX_LEN, VOCAB_SIZE]
+        """
+        # encoding
+        row_enc_out, hidden, context = self.encode(features)
+        # init decoder's states
+        h, c, O_t = self.init_decoder(row_enc_out, hidden, context)
+
+        if formulas is not None:
+            logits, targets = self.decode_with_formulas(
+                h, c, O_t, formulas, row_enc_out)
+        elif self.beam_width == 0:
+            device = features.device
+            b_size = features.size(0)
+            logits, targets = self.decode_without_formulas(
+                h, c, O_t, row_enc_out, b_size, device)
+        else:
+            assert features.size(
+                0) == 1, "Using beam search, batch size must be equal to zero"
+            device = features.device
+            b_size = features.size(0)
+            logits, targets = self.decode_with_bs(
+                h, c, O_t, row_enc_out, b_size, device)
+
+        return logits, targets
+
+    def decode_with_formulas(self, h, c, O_t, formulas, row_enc_out):
+        logits = []
+        # do not use beam search; predict as usual
+        max_len = formulas.size(1)
+        for t in range(max_len):
+            tgt = formulas[:, t:t+1]
+            h, c, O_t, logit = self.step_decoding(
+                h, c, O_t, row_enc_out, tgt)
+            logits.append(logit)
+        logits = torch.stack(logits, dim=1)  # [B, MAX_LEN, out_size]
+        targets = torch.max(torch.log(logits).data, dim=2)[1]
+        return logits, targets
+
+    def decode_without_formulas(self, h, c, output, row_enc_out, b_size, device):
+        logits = []
+        logit = None
+        for _ in range(self.max_len):
+            if logit is not None:
+                tgt = torch.reshape(torch.max(logit.data, dim=1)[
+                                    1], (b_size, 1)).clone().detach()
+            else:
+                tgt = torch.tensor([[START_TOKEN]] * b_size)
```

Comment:
why tgt

Reply:
I mean it is not target

Reply:
The variable name is target because it is the previous decoded target. Model returns two variables: targets (which is simply a sequence of token numbers) and logits (which are target probability distributions used for calculating loss).

Reply:
Target is basically thing that should be predicted, **groundtruth**

Reply:
Ilya, in the function `decode_without_formulas` each target token is predicted.

(The name "target" is used in the [original paper](https://arxiv.org/pdf/1609.04938.pdf): `The target markup tokens {y_t} are then generated by a decoder based on a sequence of annotation grid V`)